backend/main/src/Group/Repositories/GroupRepository.ts

```diff
@@ -0,0 +1,154 @@
+			groupId: "",
+			groupName: "",
+			userIds: [],
+			containerIds: [],
+		};
+
+		if (id instanceof GroupId) {
+			const GroupDetailData = await this.nishikiDbClient.getGroup({
+				groupId: id.id,
+			});
+			if (!GroupDetailData) return null;
+			groupData.groupId = GroupDetailData.groupId;
+			groupData.groupName = GroupDetailData.groupName;
+		}
+
+		// if the ID is containerId
+		const GroupDetailData = await this.nishikiDbClient.getGroup({
+			containerId: id.id,
+		});
+		if (!GroupDetailData) return null;
+		groupData.groupId = GroupDetailData.groupId;
+		groupData.groupName = GroupDetailData.groupName;
+
+		// get list of containers and users in the group.
+		const listOfContainers = await this.nishikiDbClient.listOfContainers(
+			GroupDetailData.groupId,
+		);
+		const listOfUsersInGroup = await this.nishikiDbClient.listOfUsersInGroup(
+			GroupDetailData.groupId,
+		);
+		groupData.containerIds = listOfContainers;
+		groupData.userIds = listOfUsersInGroup.map((user) => user.userId);
+
+		return createGroupObject(groupData);
+	}
+
+	/**
+	 *
+	 * Create a Group.
+	 * @param Group
+	 */
+	async create(Group: Group): Promise<undefined> {
+		await this.nishikiDbClient.saveGroup(Group.id.id, {
+			groupName: Group.name,
+			userIds: Group.userIds.map((userId) => userId.id),
+			containerIds: Group.containerIds.map((containerId) => containerId.id),
+		});
+	}
+
+	/**
+	 * Update a Group
+	 * @param Group
+	 */
+	async update(Group: Group): Promise<undefined> {
+		await this.nishikiDbClient.saveGroup(Group.id.id, {
+			groupName: Group.name,
+			userIds: Group.userIds.map((userId) => userId.id),
+			containerIds: Group.containerIds.map((containerId) => containerId.id),
+		});
+	}
+
+	/**
+	 * Delete a Group
+	 * @param id - target Group id
+	 */
+	async delete(id: GroupId): Promise<undefined> {
+		await this.nishikiDbClient.deleteGroup(id.id);
+	}
+}
+
+/**
+ * Create a Group from primitive values.
+ * When groupData is invalid, this function throws an error.
+ * @param groupData
+ * @throws GroupRepositoryError - this error will be thrown when the data is invalid.
+ */
+const createGroupObject = (
+	groupData: GroupWithRelationsPrimitiveData,
+): Group => {
+	const groupIdOrErr = GroupId.create(groupData.groupId);
+	const groupnameOrErr = GroupName.create(groupData.groupName);
+
+	const errorResult = hasError([groupIdOrErr, groupnameOrErr]);
+	if (errorResult.err) {
+		const report = [`GroupId: ${groupData.groupId}`, errorResult.error.message];
+		throw new GroupRepositoryError(errorResult.error.message, report);
+	}
+	const groupId = groupIdOrErr.unwrap();
+	const name = groupnameOrErr.unwrap();
+
+	const groupOrError = Group.create(groupId, {
+		name: name.name,
+		userIds: groupData.userIds.map((userId) => {
+			const userIdOrError = UserId.create(userId);
+			if (userIdOrError.ok) return userIdOrError.unwrap();
+			throw new GroupRepositoryError(userIdOrError.error.message, []);
```

Comment:
Should I make a function converting the string to UserId for readability? or is this implementation fine?

Reply:
Thank you! I couldn't care about the logging. I've fixed them